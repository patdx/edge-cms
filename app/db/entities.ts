import { Type } from '@sinclair/typebox';

const defineEntities = <T extends Record<string, unknown>>(entities: T) =>
	entities;

export const defaultEntities = defineEntities({
	categories: Type.Object(
		{
			id: Type.Optional(Type.Integer()),
			name: Type.String(),
			description: Type.Optional(
				Type.String({
					'ui:widget': 'textarea',
					'ui:options': {
						rows: 4,
					},
				}),
			),
		},
		{
			title: 'categories',
		},
	),
	posts: Type.Object(
		{
			id: Type.Optional(Type.Integer()),
			title: Type.String(),
			text: Type.Optional(
				Type.String({
					'ui:widget': 'textarea',
					'ui:options': {
						rows: 4,
					},
				}),
			),
		},
		{
			title: 'posts',
		},
	),
});
