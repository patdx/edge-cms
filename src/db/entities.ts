import { Type } from '@sinclair/typebox';

const defineEntities = <T extends Record<string, unknown>>(entities: T) =>
  entities;

export const defaultEntities = defineEntities({
  categories: Type.Object(
    {
      id: Type.Optional(Type.Integer()),
      name: Type.String(),
    },
    {
      title: 'categories',
    }
  ),
  posts: Type.Object(
    {
      id: Type.Optional(Type.Integer()),
      title: Type.String(),
      text: Type.String(),
    },
    {
      title: 'posts',
    }
  ),
});
