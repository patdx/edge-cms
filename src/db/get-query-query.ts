import type { JSONSchema7, JSONSchema7TypeName } from "json-schema";
import { ENTITY_MAP } from "./entities";
import { format } from "sql-formatter";

const escapeIdIfNeeded = (text: string) => {
  // no-op for now
  return text;
};

const getSqlType = (
  typeName: JSONSchema7TypeName
): "INTEGER" | "REAL" | "TEXT" | "BLOB" => {
  if (typeName === "string") {
    return "TEXT";
  } else if (typeName === "integer") {
    return "INTEGER";
  } else if (typeName === "number") {
    return "REAL";
  } else {
    return typeName.toUpperCase() as any;
  }
};

const getColumnDef = (
  name: string,
  schema: JSONSchema7,
  parent?: JSONSchema7
) => {
  return [
    escapeIdIfNeeded(name),
    getSqlType(schema.type),
    ...(name === "id" ? ["PRIMARY KEY"] : []),
    ...(parent?.required?.includes(name) ? ["NOT NULL"] : []),
  ].join(" ");
};

export const getCreateTableQuery = (
  schema: JSONSchema7,
  options?: {
    name: string;
    format?: boolean;
  }
) => {
  const name = options?.name ?? schema.title;

  if (!name) throw new Error(`Name of table is required`);

  const query = `CREATE TABLE ${escapeIdIfNeeded(name)} (${Object.entries(
    schema.properties ?? {}
  )
    .map(([name, options]) => getColumnDef(name, options as any, schema))
    .join(", ")}) STRICT;`;

  if (options?.format) {
    return format(query, {
      language: "sqlite",
      keywordCase: "upper",
    });
  } else {
    return query;
  }
};

export const getAllCreateQueries = (options?: { format?: boolean }) => {
  const lines: string[] = [];

  for (const [name, entity] of Object.entries(ENTITY_MAP)) {
    const query = getCreateTableQuery(entity, {
      name,
      format: options?.format,
    });

    lines.push(query);
  }

  return lines;
};

console.log(getAllCreateQueries());
