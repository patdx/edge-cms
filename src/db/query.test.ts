import { describe, expect, it } from "vitest";
import { Post } from "./entities";
import { getCreateTableQuery } from "./get-query-query";

describe("can create table", () => {
  it("post table", () => {
    expect(
      getCreateTableQuery(Post, {
        name: "Post",
      })
    ).toMatchInlineSnapshot(
      '"CREATE TABLE Post (id INTEGER PRIMARY KEY, title TEXT NOT NULL, text TEXT NOT NULL) STRICT;"'
    );
  });

  // it("match formatting")
});
