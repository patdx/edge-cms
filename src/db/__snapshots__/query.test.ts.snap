// Vitest Snapshot v1

exports[`can create table > post table 1`] = `"CREATE TABLE [object Object] (id INTEGER PRIMARY KEY, title TEXT NOT NULL, text TEXT NOT NULL) STRICT;"`;
